### DISCUSSION TITLE
Best way to handle complex permissions creation/migration

### DISCUSSION THREAD

**Discussion by 8byr0** [directus/directus#9542] [Archive]
Hey there !
I'm currently migrating a huge app from strapi to directus and I'm struggling a bit about the permissions.
In facts I have more than 30 content types with many relations from one to others and I need to setup a robust permissions model.
I love the way RBAC is doing when using directus but:
- I'm a developer, I do not want to spend a day clicking on buttons in directus' UI
- I have three environments (dev, staging, prod) so I need to be able to migrate permissions from one to another
- this is a multi-tenancy app and I'm not sure yet if I'll be using a single db for all tenants or use on demand instances of directus
- I deploy directus w/ a Ci/CD pipeline, the best is when nothing has to be done manually
- I need these permissions to be versionned somehow

Anyway, I'm not here to complain, I have  a few approches to address these requirements but I'd be glad to have your opinion for the most appropriate way to manage it:
1. Use custom migrations (docker runtime) ensuring that all is properly setup
2. Use JS SDK to manipulate relations (same as before but using TS). Issue with this is that I cannot to it fluently in my pipeline so I need to way for the container to be up and running; or bundle a script for runtime command
3. Deploy a configuration file just like for db scheme (but not possible yet from what I've seen so far)

Final point, I'd prefer not having to write SQL queries. I can do it but it's not my preferred stack.
Any thoughts ?

**Comment by benhaynes** [comment]
> I'm a developer, I do not want to spend a day clicking on buttons in directus' UI

While it doesn't answer/solve _all_ your questions, you can click on the permission's field label to get access to the raw JSON value. That might be one way to bypass the UI and create rules in code, then paste them in. 👍 

Appreciate the insight and feedback!

**Reply by 8byr0** [reply]
Yes, already found this one very useful ! But having to do it for such amount of data is a pain... Maybe a solution could be to be able to manage the whole permissions as JSON! 

But as stated in my post this is not a feature proposal, I'm trying to find the best way with what's available now!

**Reply by benhaynes** [reply]
Absolutely! Might make sense to add "helper" buttons for copying some/all config throughout roles/permissions. Just need to think through that scope.

**Reply by 8byr0** [reply]
> Absolutely! Might make sense to add "helper" buttons for copying some/all config throughout roles/permissions. Just need to think through that scope.

Sure ! The way it's done for snapshoting schema is way enough for a decent CI/CD lead project so the same would be with permissions!

**Reply by 8byr0** [reply]
After thinking a bit you pointed out something interesting: this feature is already available since:
- It's yet possible to get all permissions as json (https://docs.directus.io/reference/system/permissions/#list-permissions) 
- It's also possible to post multiple permissions at a time (e.g. altered result of request above)

I guess it could be a workaround to manage it easily, still need to figure out how to automate it (probably with an access token dedicated to CI / docker runtime)

**Comment by 8byr0** [comment]
@benhaynes In order to make this as robust as possible I'm working on a package that'll allow to generate easily my permission objects and update remote. 

TL;DR; Would you like to have such feature in directus' SDK ?



What I plan so far is having something like this:
```typescript
// index.ts
import { Directus } from "@directus/sdk";
import {
  and,
  belongsToCompany,
  hasReadPermission,
  hasWritePermission,
  isOwner,
  or,
} from "./policies.js";

// Local
const CONFIG = {
  server: "http://localhost:8055/",
};

const directus = new Directus(CONFIG.server);

await directus.auth.static("azertyadmin");

const roles = await directus.roles.readMany();
const permissions = await directus.permissions.readMany();

const buildRolePermissions = (name) => {
  switch (name) {
    case "sysadmin":
      return {
        project: {
          read: or(isOwner, and(belongsToCompany, hasReadPermission)),
          create: or(isOwner, and(belongsToCompany, hasWritePermission)),
          update: or(isOwner, and(belongsToCompany, hasWritePermission)),
          delete: or(isOwner, and(belongsToCompany, hasWritePermission)),
        },
      };
  }
};

const sysAdminRolePermissions = buildRolePermissions("sysadmin");

// Process generated permissions to convert to a valid payload
// Update existing roles / ensure they exist using SDK
```

Where all the policies would be defined like this: 
```typescript
// policies.ts
export const and = (a, b) => {
  return { _and: [a, b] };
};

export const or = (a, b) => {
  return { _or: [a, b] };
};

/**
 * grant = user_created.id === current_user.id
 */
export const isOwner = {
  user_created: {
    _eq: "$CURRENT_USER",
  },
};

/**
 * grant = doc.company === current_user.company
 */
export const belongsToCompany = {
  company: {
    _eq: "$CURRENT_USER.company",
  },
};

export const hasReadPermission = {
  sharedReadonly: {
    _contains: "$CURRENT_USER",
  },
};
export const hasWritePermission = {
  sharedReadWrite: {
    _contains: "$CURRENT_USER",
  },
};
```

The generated permissions set would look like this:
```json
{
  "project": {
    "read": {
      "_or": [
        {
          "user_created": {
            "_eq": "$CURRENT_USER"
          }
        },
        {
          "_and": [
            {
              "company": {
                "_eq": "$CURRENT_USER.company"
              }
            },
            {
              "sharedReadonly": {
                "_contains": "$CURRENT_USER"
              }
            }
          ]
        }
      ]
    },
    "create": {
      "_or": [
        {
          "user_created": {
            "_eq": "$CURRENT_USER"
          }
        },
        {
          "_and": [
            {
              "company": {
                "_eq": "$CURRENT_USER.company"
              }
            },
            {
              "sharedReadWrite": {
                "_contains": "$CURRENT_USER"
              }
            }
          ]
        }
      ]
    },
    "update": {
      "_or": [
        {
          "user_created": {
            "_eq": "$CURRENT_USER"
          }
        },
        {
          "_and": [
            {
              "company": {
                "_eq": "$CURRENT_USER.company"
              }
            },
            {
              "sharedReadWrite": {
                "_contains": "$CURRENT_USER"
              }
            }
          ]
        }
      ]
    },
    "delete": {
      "_or": [
        {
          "user_created": {
            "_eq": "$CURRENT_USER"
          }
        },
        {
          "_and": [
            {
              "company": {
                "_eq": "$CURRENT_USER.company"
              }
            },
            {
              "sharedReadWrite": {
                "_contains": "$CURRENT_USER"
              }
            }
          ]
        }
      ]
    }
  }
}
```

This approach is really focused on combination of policies to allow an action.
Would it make sense to add such feature to directus' SDK directly ? Or is it better to keep it separated ?

**Reply by 8byr0** [reply]
One side-effect of this PBAC (Policy Based Access Control) approach is that all your users would then have to share the same role. 
This could be leveled up when multi roles feature will be available. For example a user could have two roles :
- team_members (deny all)
- base_policies_role (rules defined as shown above)

This would allow to use any deny role as a reference for custom policy (e.g. userHasRole policy which would check if user as "team_members" role).

**Reply by taco-chen** [reply]
Are you going to release these as a directus extension? I'm really looking forward to your extension since I have the same pain here.

**Comment by Joey92** [comment]
So I have also always been struggling to keep my prod and dev environment in sync when it comes to permissions.. If it could be exported to yaml just like the database schema snapshot (Or even integrate it into that), that would already cover my biggest pain with it.

**Reply by linus1703** [reply]
I'm facing the same problem and personally think, that it's a good idea to include (maybe optional flag) directus tables in the schema.

**Comment by nicolashefti** [comment]
Might be worth mentionning in 2025 the existence of [Directus Template CI](https://github.com/directus-labs/directus-template-cli) which allows you to export basically everything in JSON and apply it to any other Directus instance. Still beta, but I managed to migrate my whole permissions setup to production.
